### PR TITLE
Externalize plugins

### LINKED ISSUE #44: Plugins architecture
## tl;dr;

There's no plugin infrastructure in the code, resulting in `if (angular)` `if (jquery)`.

Create one that allows us to remove the specific builds code from the main `algoliasearch.js`.

## TODO

- [ ] create a plugin way to create specific jQuery and Angular builds rather than storing the custom code in the main `algoliasearch.js` file.
- [ ] Remove specific builds code from main algoliasearch.js

## Boring stuff

We have [jQuery and Angular.js builds](https://github.com/algolia/algoliasearch-client-js/blob/4d28d22e7726b537c160190f337e2cafd68b0d63/README.md#jquery-angularjs) but the code needed by theses plugins is stored [inside the main algoliasearch.js](https://github.com/algolia/algoliasearch-client-js/blob/4d28d22e7726b537c160190f337e2cafd68b0d63/src/algoliasearch.js#L673) file.

This is not good because once we have 10 plugins we will be screwed and it complexifies the code a lot.

We must find a way to plug theses builds

We must have one entrypoint per plugin like today: [src/algoliasearch.angular.js](https://github.com/algolia/algoliasearch-client-js/blob/4d28d22e7726b537c160190f337e2cafd68b0d63/src/algoliasearch.angular.js) but it should store his own custom code.

### PR DESCRIPTION
**Pull request by vvo**
* Externalize plugins and request implementations
* NEW FEATURE, All calls are now returning promises.
 - If there's a callback given to an API call, you won't get a promise.
 - Here are the different promises implementations we use:
     - Native promises by default (https://github.com/jakearchibald/es6-promise/)
     - AngularJS promises for AngularJS plugin
     - jQuery promises for jQuery plugin

fixes #44 #45

### CODE REVIEW

File: .zuul.yml [hunk 1]
```diff
@@ -19,6 +19,7 @@ browsers:
 server: ./test/support-server/index.js
 scripts:
   - /node_modules/jquery/dist/jquery.min.js
+  - /node_modules/jQuery-ajaxTransport-XDomainRequest/jquery.xdomainrequest.min.js
```
Comment (vvo) [inline .zuul.yml]: this allows testing the jQuery plugin in IE9, this will be added to the changelog and migration guide

**Comment by vvo** [comment]
The idea behind this PR is to externalize every plugin implementation, thus preparing the work towards #46: merging JavaScript clients.

It also remove complexity and code paths from the main file.

Finally, it provides a double callback() and promises library not just for jQuery and AngularJS plugins but also for regular users, using native promises.

**Comment by vvo** [comment]
@redox @bobylito review!?

File: src/create-algoliasearch.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// this file is a `factory of algoliasearch()`
+// Given a `request` param, it will provide you an AlgoliaSearch client
+// using this particular request
+module.exports = createAlgoliasearch;
+
+function createAlgoliasearch(request) {
+  function algoliasearch(applicationID, apiKey, opts) {
+    var AlgoliaSearch = require('./algoliasearch');
+
+    return new AlgoliaSearch(applicationID, apiKey, opts, request);
```
Comment (bobylito) [inline src/create-algoliasearch.js]: style : prefer parameters for partial application in the first place

**Pull request merged**